pkg/vsphere/guestinfo/encode.go

```diff
@@ -0,0 +1,291 @@
+	default:
+		return ""
+	}
+}
+
+// Scope returns the uint representation of scope tag
+func Scope(scopes []string) uint {
+	var scope uint
+	for _, v := range scopes {
+		switch v {
+		case "hidden":
+			scope |= Hidden
+		case "read-only":
+			scope |= ReadOnly
+		case "read-write":
+			scope |= ReadWrite
+		case "non-persistent":
+			scope |= NonPersistent
+		case "volatile":
+			scope |= Volatile
+		default:
+			return Invalid
+		}
+	}
+	return scope
+}
+
+// CalculateKey creates a new key based on the scope and prefix
+func CalculateKey(scopes []string, prefix string, key string) string {
+	scope := Scope(scopes)
+	if scope&Invalid != 0 || scope&NonPersistent != 0 {
+		return ""
+	}
+
+	var guestinfo string
+	var seperator string
+
+	// Trim the whitespaces
+	key = strings.TrimSpace(key)
+
+	if scope&Hidden != 0 {
+		guestinfo = ""
+		seperator = "~"
+	}
+	if scope&ReadOnly != 0 {
+		guestinfo = "guestinfo"
+		seperator = "/"
+	}
+	if scope&ReadWrite != 0 {
+		guestinfo = "guestinfo"
+		seperator = "."
+	}
+
+	if guestinfo == "" && prefix == "" {
+		return key
+	}
+
+	if guestinfo == "" {
+		return strings.Join([]string{prefix, key}, seperator)
+	}
+
+	if prefix == "" {
+		return strings.Join([]string{guestinfo, key}, seperator)
+	}
+
+	return strings.Join([]string{guestinfo, prefix, key}, seperator)
+}
+
+// Encode convert given type to []types.BaseOptionValue
+func Encode(i interface{}) []types.BaseOptionValue {
+	//log.SetLevel(log.DebugLevel)
+
+	return EncodeWithPrefix(i, DefaultPrefix)
+}
+
+// EncodeWithPrefix convert given type to []types.BaseOptionValue
+func EncodeWithPrefix(i interface{}, prefix string) []types.BaseOptionValue {
+	var config []types.BaseOptionValue
+
+	// value representing the run-time data
+	value := reflect.ValueOf(i)
+	log.Debugf("Value: %#v", value)
+
+	// determine the kind as it changes how we get underlying data
+	switch value.Kind() {
+	case reflect.Invalid:
+		log.Errorf("Invalid Kind: %#v", value)
+		return nil
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Bool, reflect.String, reflect.Float32, reflect.Float64:
+		// handle basic types directly
+		log.Debugf("Basic type: %s", prefix)
+
+		return append(config, &types.OptionValue{Key: prefix, Value: ToString(value)})
+	case reflect.Slice, reflect.Array:
+		log.Debugf("Slice BEGIN: %s", prefix)
+
+		var values []string
+		// iterate over slice to find what's underneath
+		for i := 0; i < value.Len(); i++ {
+			// recurse if we have a struct type
+			// otherwise add it's value to values
+			if value.Index(i).Kind() == reflect.Struct {
+				key := fmt.Sprintf("%s|%d", prefix, i)
+				config = append(config, EncodeWithPrefix(value.Index(i).Interface(), key)...)
+			} else {
+				values = append(values, ToString(value.Index(i)))
+			}
+		}
+		// set the slice key with the values seperated by |
+		if len(values) > 0 {
+			// sort the values before joining
+			sort.Strings(values)
+			// prefix~ contains the items
```

Comment:
again, not sure what the impact of using ~ is on the underlying guestinfo mechanism. ditto for |